src/geometry/manifold/algebra/left_invariant_derivation.lean

```diff
@@ -0,0 +1,183 @@
+
+instance : add_comm_group (left_invariant_derivation I G) :=
+{ add_assoc := λ X Y Z, by { ext1 f, simp only [coe_add], rw add_assoc ⇑X Y Z },
+  zero_add := λ X, by { ext1 f, simp only [coe_add, coe_zero], rw zero_add ⇑X },
+  add_zero := λ X,by { ext1 f, simp only [coe_add, coe_zero], rw add_zero ⇑X },
+  add_comm := λ X Y, by { ext1 f, simp only [coe_add], rw add_comm ⇑X Y },
+  add_left_neg := λ X, by { ext1 f, simp only [coe_add, coe_neg, coe_zero], rw add_left_neg ⇑X },
+  ..left_invariant_derivation.has_zero,
+  ..left_invariant_derivation.has_add,
+  ..left_invariant_derivation.has_neg }
+
+@[simp] lemma coe_sub : ⇑(X - Y) = X - Y := by simp only [sub_eq_add_neg, coe_add, coe_neg]
+
+instance : module 𝕜 (left_invariant_derivation I G) :=
+module.of_core $
+{ smul := λ r X, ⟨r • X, λ f g, by { simp only [derivation.Rsmul_apply, algebra.id.smul_eq_mul,
+            mul_eq_mul_left_iff, linear_map.map_smul, left_invariant'], left, refl }⟩,
+  mul_smul := λ r s X, ext $ λ b, mul_smul _ _ _,
+  one_smul := λ X, ext $ λ b, one_smul 𝕜 _,
+  smul_add := λ r X Y, ext $ λ b, smul_add _ _ _,
+  add_smul := λ r s X, ext $ λ b, add_smul _ _ _ }
+
+@[simp] lemma map_smul : X (r • f) = r • X f := linear_map.map_smul X r f
+@[simp] lemma smul_map : (r • X) f = r • (X f) := rfl
+@[simp] lemma leibniz : X (f * f') = f • X f' + f' • X f := X.leibniz' _ _
+@[simp] lemma lift_smul (k : 𝕜) : (↑(k • X) : derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯) = k • X := rfl
+
+/-- Evaluation at a point for left invariant derivation. Same thing as for generic global
+derivations.-/
+def eval_at : (left_invariant_derivation I G) →ₗ[𝕜] (point_derivation I G g) :=
+{ to_fun := λ X, derivation.eval_at g ↑X,
+  map_add' := λ X Y, rfl,
+  map_smul' := λ k X, rfl }
+
+lemma eval_at_apply : eval_at g X f = (X f) g := rfl
+
+@[simp] lemma eval_at_coe : derivation.eval_at g ↑X = eval_at g X := rfl
+
+lemma left_invariant : (fd (Lb I G g)) (1 : G) (eval_at (1 : G) X) f = eval_at g X f :=
+(X.left_invariant'' f g)
+
+lemma eval_at_mul : eval_at (g * h) X f = (fd (Lb I G g)) h (eval_at h X) f :=
+by rw [←left_invariant, Lb_mul, ←fdifferential_comp, function.comp, apply_fdifferential,
+  left_invariant, ←apply_fdifferential]
+
+lemma comp_Lb : (X f).comp (Lb I G g) = X (f.comp (Lb I G g)) :=
+by ext h; rw [times_cont_mdiff_map.comp_apply, Lb_apply, ←eval_at_apply, eval_at_mul,
+  apply_fdifferential, eval_at_apply]
+
+instance : has_bracket (left_invariant_derivation I G) (left_invariant_derivation I G) :=
+{ bracket := λ X Y, ⟨⁅(X : derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯), Y⁆, λ f g, begin
+    have hX := left_invariant' g X (Y f), have hY := left_invariant' g Y (X f),
+    simp only [apply_fdifferential, derivation.eval_apply,
+      coe_lift_eq_coe, comp_Lb] at hX hY,
+    simp only [apply_fdifferential, derivation.eval_apply,
+      derivation.commutator_apply, coe_lift_eq_coe,
+      smooth_map.coe_sub, pi.sub_apply, hX, hY, comp_Lb], end⟩ }
```

Comment:
```suggestion
    rw [apply_fdifferential, derivation.eval_apply,
        derivation.commutator_apply, coe_lift_eq_coe, coe_lift_eq_coe,
        smooth_map.coe_sub, pi.sub_apply, hX, hY],
    refl,
    end⟩ }
```

CI is timing out here. I don't know if this is any quicker though?

Reply:
I applied a variation of that which is quick enough... all the same shouldn't the problem of `simp` looking for subsingleton instances be solved anyways?

Reply:
@gebner what about subsingleton instances? If you replace the rws above with
<!--StartFragment-->simp only [apply_fdifferential, derivation.eval_apply,      coe_lift_eq_coe, comp_Lb] at hX hY,    simp only [apply_fdifferential, derivation.eval_apply,      derivation.commutator_apply, coe_lift_eq_coe,      smooth_map.coe_sub, pi.sub_apply, hX, hY, comp_Lb]<!--EndFragment-->
the proofs takes 50s and it triggers a timeout (or at least it did before, but it still takes long, I verified) and my impression was that it was because lean was looking for some random subsingleton instance and it was producing ~14Mb outputs when looking for it... With the above proof it takes ~25s and it does not trigger a time out, so it is not necessary to take care of this, but maybe it is important anyways

Reply:
Yes, this is a well-known problem.